bam_qc_metrics/bam_qc.py

```diff
@@ -15,103 +51,73 @@ class bam_qc:
         'run name',
         'sample'
     ]
-    PRECISION = 1 # number of decimal places for rounded output
-    FINE_PRECISION = 3 # finer precision, eg. for reads_per_start_point output
     RANDOM_SEED = 42
-    READ_1_INDEX = 0
-    READ_2_INDEX = 1
-    READ_UNKNOWN_INDEX = 2
-    START_POINTS_KEY = 'start points'
-
-    def __init__(self, bam_path, target_path, expected_insert_max, metadata_path=None,
-                 mark_duplicates_path=None, n_as_mismatch=False, skip_below_mapq=None,
-                 reference=None, sample_rate=None, tmpdir=None, verbose=True):
-        # define instance variables
-        self.bedtools_metrics = None
-        self.custom_metrics = None
-        self.expected_insert_max = expected_insert_max
-        self.mark_duplicates_metrics = {
-            "ESTIMATED_LIBRARY_SIZE": None,
-            "HISTOGRAM": {},
-            "LIBRARY": None,
-            "PERCENT_DUPLICATION": None,
-            "READ_PAIRS_EXAMINED": None,
-            "READ_PAIR_DUPLICATES": None,
-            "READ_PAIR_OPTICAL_DUPLICATES": None,
-            "UNMAPPED_READS": None,
-            "UNPAIRED_READS_EXAMINED": None,
-            "UNPAIRED_READ_DUPLICATES": None
-        }
-        self.metadata = None
-        self.mismatches_by_read = None
-        self.qc_input_bam_path = None
+
+    def __init__(self, config):
+        self.validate_config_fields(config)
```

Comment:
This might be a matter of preference. I think there is some code duplication happening in this function. The key strings are defined in two places.

Class constant::
https://github.com/oicr-gsi/bam-qc-metrics/blob/9ed35d29ad0686f4b232055e0303e17fb9783afe/bam_qc_metrics/bam_qc.py#L22-L32

In args parsing:
https://github.com/oicr-gsi/bam-qc-metrics/blob/9ed35d29ad0686f4b232055e0303e17fb9783afe/bin/run_bam_qc.py#L103-L114

The function is ensuring that these two pieces of code stay in sync. One way around this is to define the key strings as a constant and then use them in two places:
```
KEY_BAM = 'bam'
CONFIG_KEY_BAM = KEY_BAM
config = { 
     KEY_BAM: args.bam,
}
```
Another way is to use the`bam_qc` class as the source of the string:

```
config = { 
     bam_qc.CONFIG_KEY_BAM: args.bam,
}

Reply:
Yes, that is an oversight. I think the second way is best to resolve it.

Reply:
Now fixed. :-)